### PR TITLE
New homogeneous compositions.

### PR DESCRIPTION
**Pull request by favonia**
This pull thread is to implement the new homogeneous compositions for the **Ultimate Dream.**

Current goals:

- [x] Implement the optimized adjacency conditions.
- [x] Implement the even more optimized adjacency conditions.
- [x] Get #205 merged.
- [x] Check the equations in tactic.

Current non-goals in this pull request:

- Give `com` more love.
- Implement `ghcom`.
- New machine stepper to avoid redundant checking. Related to #192.

### CODE REVIEW

**Comment by favonia** [comment]
@jonsterling It is quite amazing that the code became shorter as well. I think we made a theoretically and practically good choice.

**Comment by jonsterling** [comment]
This is beginning to look cool!

I wonder if later on we should also consider changing the concrete notation of hcoms to look a bit more like Swedish systems… Something like `hcom{r ~> r'}(A) [(i = 1) -> ...]`. No need to worry about it for now, but it might make programming a little nicer.

**Comment by favonia** [comment]
@jonsterling Right now, all the parameters belong to the operator, right? Unless you want to change this, the equation/tube-separation is probably unavoidable.

**Comment by jonsterling** [comment]
@favonia I'm talking about *concrete notation* / what we parse & print. We can arrange the notation in any way we please, even though internally there will be this separation between parameters and arguments.

File: src/redprl/operator.sml [hunk 1]
```diff
@@ -482,8 +487,11 @@ struct
     fun spanToString f (r, r') =
       P.toString f r ^ " ~> " ^ P.toString f r'
 
-    fun extentsToString f =
-      ListSpine.pretty (P.toString f) ","
+    fun equationToString f (r, r') =
+      P.toString f r ^ "=" ^ P.toString f r'
```
Comment (favonia) [inline src/redprl/operator.sml]: @jonsterling [extremely minor isssue] Do you want spaces before and after the equal sign?

**Comment by favonia** [comment]
@jonsterling Ah, I thought you were talking about hacking RedPRL not hacking _in_ RedPRL. I like the "systematic" syntax. Time to convince @cangiuli.

File: src/redprl/operator.sml (hunk 1 above)
Comment (jonsterling) [inline src/redprl/operator.sml]: @favonia I don't really have a preference, so you can write it as you prefer.

**Comment by favonia** [comment]
@jonsterling I would like to get this merged before trying to simplify the machine stack semantics.

**Comment by favonia** [comment]
@jonsterling Maybe we can merge the current progress first.

**Comment by favonia** [comment]
@jonsterling I am waiting for the ORDERED, but you can also just merge 5e13ec93943c0d9db58513f1961a2db8acc24a56.

**Comment by favonia** [comment]
@jonsterling Actually, let me separate these into two pull requests.

File: src/redprl/syntax.sml [hunk 2]
```diff
@@ -25,7 +26,7 @@ struct
    (* path: path abstraction and path application *)
    | PATH_TY of (symbol * 'a) * 'a * 'a | PATH_ABS of symbol * 'a | PATH_AP of 'a * param
    (* hcom *)
-   | HCOM of param list (* extents *) * dir * 'a (* type *) * 'a (* cap *) * (symbol * 'a) list (* tubes *)
+   | HCOM of {dir: dir, class: 'a, cap: 'a, tubes: (equation * (symbol * 'a)) list}
```
Comment (jonsterling) [inline src/redprl/syntax.sml]: good.

**Pull request merged**